runtime/src/main/java/org/corfudb/runtime/object/MVOCache.java

```diff
@@ -0,0 +1,279 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+/**
+ * MVOCache is the centralized container that holds the reference to
+ * all MVOs. Overall it provides put and get APIs and manages the
+ * Cache-related properties (LRU) under the hood.
+ *
+ */
+@Slf4j
+public class MVOCache<T extends ICorfuSMR<T>> {
+
+    private final CorfuRuntime runtime;
+
+    // A registry to keep track of all opened MVOs
+    @Getter
+    private final ConcurrentHashMap<UUID, MultiVersionObject<T>> allMVOs = new ConcurrentHashMap<>();
+
+    // The objectCache holds the strong references to all versioned objects, and snapshotProxies
+    private final Cache<VersionedObjectIdentifier, MVOCacheEntry> objectCache;
+
+    // This objectVersions is updated at two places
+    // 1) put() which adds a new version to the objectVersions
+    // 2) as a side effect of put(), some old versions are evicted from the cache
+    // => access/mutation to objectMap should be synchronized
+    // Q: Can we use objectCache.asMap() instead of maintaining an external view?
+    // A: No. Although asMap() returns a thread-safe weakly-consistent map, it is
+    //    not a tree structure so it's very inefficient to implement headMap() and
+    //    floorEntry() on top of it.
+    @Getter
+    private final ConcurrentHashMap<UUID, TreeSet<Long>> objectVersions = new ConcurrentHashMap<>();
+
+    /**
+     * TODO: access pattern
+     * Mutations: 1) objectCache, 2) objectVersions
+     * Accesses/Gets: 1) objectVersions, 2) query objectCache for real object if previous query is true
+     */
+
+    private static final long DEFAULT_CACHE_EXPIRY_TIME_IN_SECONDS = 300;
+
+    private final ScheduledExecutorService mvoCacheSyncThread = Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder().setDaemon(true)
+                    .setNameFormat("MVOCacheSyncThread")
+                    .build());
+
+    @Getter
+    private final MVOCacheEviction mvoCacheEviction = new MVOCacheEviction(this::prefixEvict);
+
+    public MVOCache(CorfuRuntime corfuRuntime) {
+        runtime = corfuRuntime;
+
+        CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder();
+        objectCache = cacheBuilder.maximumSize(runtime.getParameters().getMaxCacheEntries())
+                .expireAfterAccess(DEFAULT_CACHE_EXPIRY_TIME_IN_SECONDS, TimeUnit.SECONDS)
+                .expireAfterWrite(DEFAULT_CACHE_EXPIRY_TIME_IN_SECONDS, TimeUnit.SECONDS)
+                .removalListener(this::handleEviction)
+                .recordStats()
+                .build();
+
+        mvoCacheSyncThread.scheduleAtFixedRate(this::syncMVOCache,
+                runtime.getParameters().getMvoAutoSyncPeriod().toMillis(),
+                runtime.getParameters().getMvoAutoSyncPeriod().toMillis(),
+                TimeUnit.MILLISECONDS);
```

Comment:
I think the default should be reduced and the logic should take account for syncs that were triggered by application threads. If this scheduler triggers a task for an object that was just sync'd it should't be resync'd again.